sklego/meta/grouped_predictor.py

```diff
@@ -397,15 +208,160 @@ def decision_function(self, X):
-        check_is_fitted(self, ["estimators_", "groups_", "fallback_"])
+        return self.__predict_estimators(X, method_name="decision_function")
+
+    def __validate_inputs(self, frame):
+        """Validate the input arrays"""
+
+        if frame.shape[1] <= len(self.groups_) + 1:
+            raise ValueError("`X` contains no features")
 
-        X_group, X_value = _split_groups_and_values(
-            X, self.groups, min_value_cols=0, check_X=self.check_X, **self._check_kwargs
-        )
+        if self.check_X:
+            X_values = frame.drop(columns=self.groups_ + ["__target_value__"]).copy()
+            check_array(X_values, **self._check_kwargs)
 
-        X_group = self.__add_shrinkage_column(X_group)
+        X_groups = frame.loc[:, self.groups_].copy()
+
+        X_group_num = X_groups.select_dtypes(include="number")
+        if X_group_num.shape[1]:
+            check_array(X_group_num, **self._check_kwargs)
+
+        # Only check missingness in object columns
+        if X_groups.select_dtypes(exclude="number").isnull().any(axis=None):
+            raise ValueError("Group columns contain NaN values")
+
+        return self
+
+    def __set_fit_levels(self):
+        """Based on the combination of parameters passed to the class, it defines the groups/levels that were fitted.
+
+        This function should be called only after assigning self.groups_ during fit.
+        """
+        check_is_fitted(self, ["groups_"])
+
+        if self.fallback_method == "raise":
+            levels_ = self.groups_ if self.shrinkage is None else expanding_list(self.groups_)
+
+        elif self.fallback_method == "next":
+            levels_ = expanding_list(self.groups_)
+
+        elif self.fallback_method == "global":
+            if not self.use_global_model:
+                raise ValueError("`fallback_method`='global' requires `use_global_model=True`")
+            elif self.shrinkage is None:
+                levels_ = [["__global_model__"], self.groups_]
+            else:
+                levels_ = expanding_list(self.groups_)
 
-        if self.shrinkage is None:
-            return self.__predict_groups(X_group, X_value, method="decision_function")
         else:
-            return self.__predict_shrinkage_groups(X_group, X_value, method="decision_function")
+            raise ValueError(f"`fallback_method` should be one of {self._ALLOWED_FALLBACK}, not {self.fallback_method}")
+
+        return levels_
+
+    def __fit_estimators(self, frame):
+        """Fit one estimator per level of the group column(s)"""
+        estimators_ = {}
+        for grp_names in self.fitted_levels_:
+            for grp_values, grp_frame in frame.groupby(grp_names):
+                _X = grp_frame.drop(columns=self.groups_ + ["__target_value__"])
+                _y = grp_frame["__target_value__"]
+
+                if _y.isnull().all():
+                    estimators_[grp_values] = clone(self.estimator).fit(_X)  # TODO: or .fit(_X, None) ?
```

Comment:
Not sure what the convention is for outlier detectors. In scikit-learn and pyod it seems that we can pass `y` and it will just get ignored anyway.
If we stand to that convention we don't need to distinguish the two cases